Libraries/LibC/stdlib.cpp

```diff
@@ -469,6 +470,28 @@ int wctomb(char*, wchar_t)
     ASSERT_NOT_REACHED();
 }
 
+size_t wcstombs(char* dest, const wchar_t* src, size_t max)
+{
+    char* originalDest = dest;
+    while ((size_t)(dest - originalDest) < max) {
+        StringView v { (const char*)src, sizeof(wchar_t) };
```

Comment:
This is wrong. `Utf8CodepointIterator` (that you create from this `StringView` below via a `Utf8View`) assumes that the string is valid UTF-8, and will panic (or could exhibit undefined behavior, though it shouldn't) otherwise. It's true that the current implementation will only panic once it stumbles upon an invalid byte sequence, so everything will appear to work if you break out of a loop in the middle, but it's still UB to pass an invalid UTF-8 to `Utf8CodepointIterator`. Don't do that.

Reply:
We should really have an API that can fail somehow productively instead of asserting that things are always valid :)

Reply:
`bool Utf8View::validate() const` is that API. It's legal to create a `Utf8View` with any blob and call `validate()` on it; it's UB to create a `Utf8CodepointIterator` out of it if it's not actually valid UTF-8.

Reply:
Moreover, it looks like this implementation is actually wrong. I believe `src` is supposed to be in `wchar_t`s (so something like UTF-16 or UTF-32), and it's supposed to be *encoded* into UTF-8 that gets written into `dest`, not the other way around. (The name `wcstombs` reads "wide characters to multi-byte string").

Reply:
Hi @bugaevc ,

I tested this function with the Python3 Port and it seemed "to work". But I'm no expert on that multibyte string thingy and tried to understand the C++ reference pages on that.

So your assumption is right, this method *should* convert a `wchar_t` array to an char* array containing a multi byte string. So my try was to use the string view of a single `wchar_t` to extract it's utf8 codepoint and then to copy it to dest.

The `validate()` is a really good point. I'll try to fix this later.

If you have any good idea how I can improve that, I'm more than happy if you let me know!

Cheers,
Emanuel

Reply:
> So my try was to use the string view of a single `wchar_t` to extract it's utf8 codepoint and then to copy it to dest.

To expand on my point, `Utf8View` can *decode* UTF-8, that is essentially convert from multi-byte to wide characters (`u32`), so what that `mbstowcs()` function does except `Utf8CodepointIterator` works as an iterator instead of converting the whole string into the caller-provided buffer in one go.

`wcstombs()` should do the reverse operation, UTF-8 *encoding*. `Utf8View` & `Utf8CodepointIterator` are of no use for that, at least in their current shape. I'm not sure if it would make more sense to add an UTF-8 encoding utility to AK (and what would be the best API for that) and wrap that or implement it right here.

Reply:
Ok thanks for your explanation. I'll dive in this topic and rewrite the function. As this pull request is already closed, I'll do that in a new pull-request, right? (Sorry I'm new to github workflow, worked with gerrit workflow previously).

Reply:
Right. And please cc/mention me (like this: cc @bugaevc) in that new PR so I'll be able to notice and review it :)

Reply:
Yeah, just open a new PR with new changes, and we can discuss them there. I prefer to not get caught up in too much discussion, but instead to merge and iterate rapidly :)

Reply:
Ok, I'll do! Thank you both :)